Documentation/AdditionalFunctionality/PushNotifications.md

```diff
@@ -0,0 +1,164 @@
+# Push Notifications
+
+The `NotificationCenter` can handle the registration and sending of **push notifications** to Apple Push Notification Service (APNS) and Firebase Cloud Messaging (FCM). APNS can send out push notifications to iOS, macOS and tvOS devices, while FCM handles messaging for iOS, Android and Web Push.
+
+## Configuration
+
+In order to send push notifications, the Apodini web server needs to be authenticated to at least one push notification service. These services are added to a web server by defining them in the `configuration` computed property.
+
+APNS can be configured using either a `.p8` or a `.pem` certificate. FCM is added using the _service-account.json_ file generated by the Firebase Console.
+
+This example adds APNS using a `.pem` certificate and FCM to the web server.
+
+```swift
+var configuration: Configuration {
+    APNSConfiguration(.pem(pemPath: "certificate.pem"), identifier: "de.tum.in.ase.Example", environment: .sandbox)
+    FCMConfiguration("firebase-adminsdk.json")
+}
+```
+
+The `NotificationCenter` can also be used to register devices. This is useful to retrieve saved devices and send out push notifications to them. This feature requires a connected **[database](./Database.md)** to Apodini. Using the special modifier `addNotifications()` with the `DatabaseConfiguration` will create a new database table called `NotificationDevice` of type `Device`.
+
+The following example configures a MongoDB database and enables the `NotificationCenter` to save devices to the _apodini_db_ database.
+
+```swift
+var configuration: Configuration {
+    // ...
+    DatabaseConfiguration(.mongo)
+        .connectionString("mongodb://localhost:27017/apodini_db")
+        .addNotifications()
+}
+```
+
+## Usage
+
+The `NotificationCenter` can be injected to any `Component` by using the `@Environment` property wrapper.
+
+```swift
+struct NewsAlertComponent: Component {
+    @Environment(\.notificationCenter) var notificationCenter: NotificationCener
+
+    // ...
+}
+```
+
+## Device Registration
+
+Devices can be registered to the `NotificationCenter` by conforming to the `Device` protocol. This protocol consists of the following properties:
+
+- **user**: The associated user of the device.
+- **type**: This property specifies which push notification service to use.
+- **token**: The device id for one app used by the push notification service
+- **subscriptions**: Subscriptions are used to group `Device`s together. Example: All devices that want to receive news alerts for a specific topic.
+
+```swift
+public struct Device {
+    public var user: User
+    public var type: PushNotificationService
+    public var token: String
+    public var subscriptions: [String]?
```

Comment:
I would maybe think if we can make the subscriptions array a bit more type safe then just a `String` identifying a topic. I don't think this has to be part of the first implementation but would be something that we might want to documental as part of the proposal.

Reply:
I'm not really sure how this would look like. FCM comes with the ability to send push notifications to topics/subscriptions which is String based. So for all devices which use FCM I would just use the functionality of FCM and for APNS devices I would individually send messages to the ids/tokens of devices.

Reply:
I see, so adding subscriptions to an APNS device would have no effect? Wouldn't it be possible to manually search for all APNS devices that include the subscription and manually send out batches of push notifications?
Regarding the String topics/subscriptions: I see the point that we will eventually map the topic to a string describing the subscription. Nevertheless I would try to avoid having string identifies across an Apodini service. E.g. if you would change the name of a subscription it would require you to manually find all occurrences of the String in the application. Maybe a key path based approach or a wrapper object that is globally accessible and is a single source of truth your be nice.

Reply:
> I see, so adding subscriptions to an APNS device would have no effect? Wouldn't it be possible to manually search for all APNS devices that include the subscription and manually send out batches of push notifications?

I think that's exactly what @awoc meant with his last sentence.

Regarding the string topic: I totally agree with @PSchmiedmayer that the Apodini API should enforce usage of an approach where the existence of topics is checked at compile-time by default. However, there might also be scenarios where the developer wants to create subscriptions at runtime (e.g. one subscription for each PR on a repository). I would suggest creating a `Subscription` protocol that requires a `identifier: String` computed property. The user could then extend `String` to implement it or use an enum e.g.:

```swift
enum Subscriptions {
    case all
    case PR(RepoId, PRId)
}

extension Subscriptions: Subscription {
    var identifier: String {
        switch self {
        case .all:
              return "all"
        case .PR(let repoId, let prId):
              return "pr:\(prId):\(repoId)"
        }
    }
}

```